src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -0,0 +1,213 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Completion;
+using Microsoft.VisualStudio.Editor.Razor;
+using SyntaxNode = Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
+{
+    internal class DefaultRazorSemanticTokenInfoService : RazorSemanticTokenInfoService
+    {
+        private readonly TagHelperFactsService _tagHelperFactsService;
+        private readonly TagHelperDescriptionFactory _tagHelperDescriptionFactory;
+        private readonly HtmlFactsService _htmlFactsService;
+        private readonly SemanticTokenLegend _symanticTokenLegend;
+
+        [ImportingConstructor]
+        public DefaultRazorSemanticTokenInfoService(
+            TagHelperFactsService tagHelperFactsService,
+            TagHelperDescriptionFactory tagHelperDescriptionFactory,
+            HtmlFactsService htmlFactsService)
+        {
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            if (tagHelperDescriptionFactory is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperDescriptionFactory));
+            }
+
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _tagHelperFactsService = tagHelperFactsService;
+            _tagHelperDescriptionFactory = tagHelperDescriptionFactory;
+            _htmlFactsService = htmlFactsService;
+
+            _symanticTokenLegend = new SemanticTokenLegend();
+        }
+
+        public override SemanticTokens GetSemanticTokens(RazorCodeDocument codeDocument, SourceLocation? location = null)
+        {
+            if (codeDocument is null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+            var syntaxTree = codeDocument.GetSyntaxTree();
+
+            var syntaxTokens = VisitAllNodes(syntaxTree);
+
+            var semanticTokens =  ConvertSyntaxTokensToSemanticTokens(syntaxTokens, codeDocument, _symanticTokenLegend);
+
+            return semanticTokens;
+        }
+
+        private static IEnumerable<SyntaxNode> VisitAllNodes(RazorSyntaxTree syntaxTree)
+        {
+            return VisitNode(syntaxTree.Root);
+        }
+
+        private static IReadOnlyList<SyntaxNode> VisitNode(SyntaxNode syntaxNode)
```

Comment:
@ajaybhargavb we have visitors for this dont we?

Reply:
Yup. Write a visitor that implements `SyntaxWalker`. Here is an example, https://github.com/dotnet/aspnetcore-tooling/blob/478e2ceda86cf1bd41a3886f42387d161f43617b/src/Razor/src/Microsoft.AspNetCore.Razor.Language/TagHelperSpanVisitor.cs#L10.